### ISSUE TITLE
Allow constructor calls for side effects with no-unused-expression, move checking to new "no-new" rule

### ISSUE THREAD

**Issue by verticalpalette** [palantir/tslint#347]
```ts
new SyncManager(fakeWebSocket, "session");
```

triggers the `no-unused-expression` rule, even though constructors are also function calls. This is difficult in testing when we're testing to make sure our class is calling the expected methods on `fakeWebSocket`.

**Comment by adidahiya** [comment]
It feels like bad style to call a constructor only for side effects and to not use the result of the function call.

**Comment by verticalpalette** [comment]
This is in tests, to make sure that during the construction of the object, that we're sending the correct messages up and down the fake websocket. If we assign to a variable, then we trigger a different lint rule (unused variable).

Our alternatives right now are to:
- disable this lint rule around each constructor call (annoying)
- disable this lint rule altogether (scary)
- assign the result to a variable, and assign something else to it, to avoid the other lint rule (gross)

I don't disagree that this (calling a constructor for side-effects) would be pretty gross in non-test code, but maybe there should be a different option for that, as opposed to rolling it up into `no-unused-expression`? I was pretty surprised to see that naked function calls were linted differently than constructor calls in this case.

**Comment by adidahiya** [comment]
an alternative would be to refactor your `SyncManager` class to do its testable stuff in a public method. so instead of

```ts
class SyncManager {
  constructor(webSocket: WebSocket, ...) {
    // do things with webSocket
  }
}
```

you might do:

```ts
class SyncManager {
  private WebSocket: webSocket;

  constructor(webSocket: WebSocket, ...) {
    this.webSocket = webSocket;
  }

  public initializeWebSocket() {
    // do things with this.webSocket
  }
}
```

IMO that'd be better code style... but that might cause you to have to do a large refactor in your code base. So we might consider removing constructor call checks from the `no-unused-expression` rule and instead making a new rule with a name like `assign-constructor-result` which you could disable in these tests. Thoughts @ashwinr, @gscshoyru?

**Comment by verticalpalette** [comment]
I really like the idea of the `assign-constructor-result` rule; it would handle this case quite cleanly.

BTW I've found another instance of this, also in tests, which occurs when I'd like to make sure a constructor will throw an error if given incorrect arguments.

```ts
assert.throws(() => {
  new MyClass(null); // throws because you must pass it a non-null argument.
});
```

In this case, it seems like a shame to have to make and call a `validateConstructorArguments` method.

**Comment by adidahiya** [comment]
it should probably be called `no-new` to align with the eslint rule: http://eslint.org/docs/rules/no-new

**Comment by bencoveney** [comment]
I'm working on a pull request for this Issue but am struggling with the tests. I don't suppose anyone would be able to take a look and tell me where I am going wrong?

https://github.com/bencoveney/tslint/tree/constuctorsUsed

When running the default grunt task I get the following message (as part of run:testRules):

Could not find implementations for the following rules specified in the configuration:
no-new
Try upgrading TSLint and/or ensuring that you have all necessary custom rules installed.
If TSLint was recently upgraded, you may have old rules configured which need to be cleaned up.

Is there some build step I am missing or some place I need to register the rule that I have overlooked? Thanks for any help you can give.

**Comment by adidahiya** [comment]
@bencoveney your problem is simple -- you _must_ name the file that contains a rule implementation with the suffix "Rule". In your case, this should be `noNewRule.ts`. https://github.com/palantir/tslint#writing-custom-rules
